### PR TITLE
97 replace auth2 system with appwrite

### LINKED ISSUE #97: Replace auth2 system with appWrite
**Short description:**
Replace actual auth2 system with AppWrite.

**Describe the solution you'd like**
Try the following tutorial : https://medium.com/appwrite-io/social-authentication-on-flutter-with-appwrite-7639529eb0cc

### PR DESCRIPTION
**Pull request by LucasTesnier**
# Purpose

The purpose of this pull request is to add the Auth Linker Page.

Fixes #97 

## Added features

* A Token handle have been implemented in server side (Database and API route)
* A new Flutter app page is now available -> Auth_linker
 - Possibility to Log / Logout from all Service API available in the project.
 - Google Token is now Fully available !
 - Discord one is... unfortunately not working for the for the Mobile version. I run to multiple bug and i will need to make a post on StackOverflow...
 
*Reminder, ask Monday for help about the Activity in Manifest.json which not activated.*
*A Facebook plugin is available to handle login, try to use it*

image
image
image

### CODE REVIEW

**Review by TomDUVAL-MAHE — DISMISSED** [review]
NTR

File: server/prisma/schema.prisma [hunk 1]
```diff
@@ -26,9 +26,18 @@ model User {
   createdAt                DateTime                   @default(now())
   googleId                 String?                    @unique
   facebookId               String?                    @unique
+  Token                    Token[]
   UsersHasActionsReactions UsersHasActionsReactions[]
 }
 
+model Token {
```
Comment (ZQUEMA) [inline server/prisma/schema.prisma]: Why did you create a Token model? What is its purpose?

File: server/prisma/schema.prisma [hunk 2]
```diff
@@ -26,9 +26,18 @@ model User {
   createdAt                DateTime                   @default(now())
   googleId                 String?                    @unique
   facebookId               String?                    @unique
+  Token                    Token[]
```
Comment (ZQUEMA) [inline server/prisma/schema.prisma]: Can a user have multiple tokens for each service?

File: server/server_app.js [hunk 3]
```diff
@@ -423,6 +425,62 @@ app.get(
...
+  async (req, res, next) => {
+    if (!req.user) return res.status(401).send('Invalid token')
+    try {
+      const TokenList = await database.prisma.Token.findMany({
+        include: { User: true }
+      })
+      let existingToken = null
+      TokenList.forEach(token => {
+        if (token.userId == req.user.id) existingToken = token
+      })
```
Comment (ZQUEMA) [inline server/server_app.js]: Do a query with a `where` option.

File: server/server_app.js [hunk 4]
```diff
@@ -423,6 +425,62 @@ app.get(
...
 )
 
+/**
+ * Route used for Create/Update auth token
+ * If no token storage is already linked with the user, a new one is created
+ * body.google The Google auth token (Set to '' to remove it)
+ * body.discord The Discord auth token (Set to '' to remove it)
+ * Route protected by a JWT token
+ */
+app.post(
```
Comment (ZQUEMA) [inline server/server_app.js:433]: There is an better way to do this.

You can fetch the user with its Token. Then, if it has a Token, update it, else update the user by creating a token.

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
Some changes are necessary and some questions need to be answered.

File: server/prisma/schema.prisma (hunk 1 above)
Comment (LucasTesnier) [reply server/prisma/schema.prisma]: The Goal of the Token model is to store all the AccessToken of a User.
It's useful to use the Service API as the User :)

Comment (ZQUEMA) [reply server/prisma/schema.prisma]: I understand, but this model should not be instantiated like this.

File: server/prisma/schema.prisma (hunk 2 above)
Comment (LucasTesnier) [reply server/prisma/schema.prisma]: updated

File: server/server_app.js (hunk 3 above)
Comment (LucasTesnier) [reply server/server_app.js]: updated

File: server/server_app.js (hunk 4 above)
Comment (LucasTesnier) [reply server/server_app.js:433]: updated

File: server/passport/token.js [hunk 5]
```diff
@@ -20,7 +20,7 @@ passport.use(
       const user = await database.prisma.User.findUnique({
         where: {
           id: jwt_payload.id
-        }
+        },
```
Comment (ZQUEMA) [inline server/passport/token.js]: Useless coma

**Review by ZQUEMA — APPROVED** [review]

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
One change

**Review by TomDUVAL-MAHE — APPROVED** [review]

**Review by ZQUEMA — APPROVED** [review]

**Pull request merged**